feat(CommunityPermission): Create main UI page for joining a community that requires permissions

noeliaSD opened the pull request:
Closes https://github.com/status-im/status-desktop/issues/9267

### What does the PR do
Separated in different commits:
- It creates `CommunityColumnHeaderPanel` component and updates `CommunityColumnView` to use it. Storybook new page added for that.
- It creates join permission overlay panel. Storybook new page added for that.
- It creates `JoinCommunityView` page containing blur custom background layout. It will allow to create different fake data that can be completely fake or we can get some information real from the specific community.  Storybook new page added for that.
- It adds `StatusMessage` component in storybook but just with 2 basic validations.

**NOTE:** `JoinCommunityView` is still not integrated in the app (waiting for backend). This pr only contemplates the UI part.

### Affected areas

- Left panel - **Community header** component
- New UI pages / components, not affecting the existing app behavior.

### Screenshot of functionality

1. `CommunityColumnHeaderPanel` component in storybook:

https://user-images.githubusercontent.com/97019400/216307790-f85cdac3-323e-42d1-8fb8-c9bddb1b5921.mov

2. `StatusMessage` component in storybook:

https://user-images.githubusercontent.com/97019400/216308897-7a8aece8-2baf-44c7-a578-2046d3f96794.mov

3. `JoinPermissionsOverlayPanel` component in storybook:

https://user-images.githubusercontent.com/97019400/216308218-72b1acc8-596d-4093-a18f-d9d6d39ecb0b.mov

4.  `JoinCommunityView` component in storybook:

https://user-images.githubusercontent.com/97019400/216309005-69d2062f-e4f1-425d-a946-7480f5305ccf.mov

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | b6c973ae | [#1](https://ci.status.im/job/status-desktop/job/branches/job/linux-cpp/job/PR-9405/1/) | 2023-02-02 11:24:27 | ~5 min | `linux-cpp` | [:page_facing_up:`log`](https://ci.status.im/job/status-desktop/job/branches/job/linux-cpp/job/PR-9405/1//consoleText) |

caybro commented on storybook/pages/JoinCommunityPermissionsEditor.qml:
```diff
@@ -0,0 +1,160 @@
+
+        RadioButton {
+            checked: true
+            text: qsTr("Join community")
```
No need for qsTr() in storybook

Also below in this file

caybro commented on storybook/pages/StatusMessagePage.qml:
```diff
@@ -0,0 +1,103 @@
+
+        property var messagesModel: ListModel {
+            ListElement {
+                timestamp: "1656937930"
```
`timestamp` is a `double` in the real model

caybro reviewed (approved):
Awesome work!

alexandraB99 commented on ui/app/AppLayouts/Chat/panels/communities/JoinPermissionsOverlayPanel.qml:
```diff
@@ -0,0 +1,100 @@
+    implicitHeight: column.implicitHeight + topPadding + bottomPadding
+    implicitWidth: column.implicitWidth + leftPadding + rightPadding
+    contentItem: ColumnLayout {
+        id: column
```
I believe width and height needs to specified in this case as a Layout is used (only no Layouts are automatically taking inherit the total size of their children eg Column or Row, ColunmLayout, RowLayout etc need instead a size to be set so that they can inherit to the children). Having it in contentItem, it seems that that ColumnLayout inherits the Control's dimensions which at the end are based on the ColumnLayout. Didn't any binding loop warning appear? did you try it with 5.15?

micieslak replied:
I think that it's ok bc when control has size not explicitly set, it has size of it's content. When the size is provided for the control, then the content is adjusted accordingly. That's the biggest advantage of Control.

alexandraB99 replied:
yeah but the size provided for the control here is the content's size

alexandraB99 replied:
implicitHeight: column.implicitHeight and column has no implicitHeight set (same for width)

micieslak replied:
This part:
```
    implicitHeight: column.implicitHeight + topPadding + bottomPadding
    implicitWidth: column.implicitWidth + leftPadding + rightPadding
```
is not needed here. Implicits are calculated according to that formula by default. I believe removing it clears all doubt here.

noeliaSD replied:
Ok, I can remove these 2 lines so they are really redundant.

alexandraB99 reviewed (commented):
just one comment, otherwise massive work!

micieslak commented on storybook/pages/CommunityColumnHeaderPanelPage.qml:
```diff
@@ -0,0 +1,101 @@
+    SplitView {
+        orientation: Qt.Vertical
+        SplitView.fillWidth: true
+
+
```
please remove empty lines

micieslak commented on storybook/pages/CommunityColumnHeaderPanelPage.qml:
```diff
@@ -0,0 +1,101 @@
+                CommunityColumnHeaderPanel {
+                    id: communityColumnHeader
+                    width: widthSlider.value
+                     anchors.centerIn: parent
```
extra space here, please add empty line after id above

micieslak commented on storybook/pages/JoinCommunityViewPage.qml:
```diff
@@ -0,0 +1,211 @@
+            color: Theme.palette.statusAppLayout.rightPanelBackgroundColor
+            clip: true
+
+            JoinCommunityView{
```
missin space

micieslak commented on storybook/src/Models/PermissionsModel.qml:
```diff
@@ -30,55 +33,196 @@ ListModel {
-                imageSource: ModelsData.collectibles.kitty1
-            }
-        ]
+        var holdings = []
```
const

micieslak commented on storybook/src/Models/PermissionsModel.qml:
```diff
@@ -30,55 +33,196 @@ ListModel {
-                imageSource: ModelsData.assets.amp
-            }
-        ]
+        var holdings = []
```
`const`. Why removing directly was wrong here? :D

micieslak replied:
Same below, please consider removing `;`s for consistency.

noeliaSD replied:
> `const`. Why removing directly was wrong here? :D

Just rebase issue. I will keep the return directly way!

micieslak replied:
> `const`. Why removing directly was wrong here? :D

sorry, removing -> returning

micieslak reviewed (commented):
Looks very good! This is my first part of comments (mainly small style issues), I will provide rest of comments in a separate review.

micieslak commented on ui/app/AppLayouts/Chat/panels/communities/JoinPermissionsOverlayPanel.qml:
```diff
@@ -0,0 +1,100 @@
+        spacing: root.spacing
+
+        HoldingsListPanel {
+            Layout.alignment: Qt.AlignHCenter
```
To properly adapt for wider content or narrow window, it should be `Layout.fillWidth: true`. Then component should center content on it's own and wrap text if necessary as well.

micieslak commented on ui/app/AppLayouts/Chat/panels/communities/HoldingsListPanel.qml:
```diff
@@ -0,0 +1,95 @@
+        spacing: root.spacing
+
+        StatusBaseText {
+            Layout.alignment: Qt.AlignHCenter
```
it should be `Layout.fillWidth` + text wrapping + `horizontalAlignment: Text.AlignHCenter` to adapt properly when window size is limited or text is long.

micieslak commented on ui/app/AppLayouts/Chat/panels/communities/HoldingsListPanel.qml:
```diff
@@ -0,0 +1,95 @@
+            spacing: d.defaultHoldingsSpacing
+
+            Repeater {
+                id: repeater
```
id not needed here

micieslak commented on ui/app/AppLayouts/Chat/panels/communities/HoldingsListPanel.qml:
```diff
@@ -0,0 +1,95 @@
+
+                        Repeater {
+                            model: SortFilterProxyModel {
+                                sourceModel: holdingsListModel
```
what is `holdingsListModel` in this context?

noeliaSD replied:
`HoldingsListPanel` has a model that represents a permission.
A permission consists on a list of different properties, one of them is `holdings list`. At this point, the repeater is working with / representing this `holdings` list model.

micieslak commented on ui/app/AppLayouts/Chat/panels/communities/HoldingsListPanel.qml:
```diff
@@ -0,0 +1,95 @@
+                    Layout.alignment: Qt.AlignHCenter
+                    spacing: d.defaultHoldingsSpacing
+
+                    RowLayout {
```
This row doesn't take available width into account. With many items/narrow window it will go offscreen.

noeliaSD replied:
It will not go offscreen since the parent has a scrollview!

micieslak commented on ui/app/AppLayouts/Chat/panels/communities/JoinPermissionsOverlayPanel.qml:
```diff
@@ -0,0 +1,100 @@
+import AppLayouts.Chat.helpers 1.0
+import AppLayouts.Chat.controls.community 1.0
+
+import utils 1.0
```
not used (?)

micieslak commented on storybook/pages/JoinCommunityViewPage.qml:
```diff
@@ -0,0 +1,211 @@
+            ListElement { name: "blockchain"; selected: true; notificationsCount: 3; hasUnreadMessages: true}
+            ListElement { name: "faq"; selected: false; notificationsCount: 0; hasUnreadMessages: false}
+        }
+        property var messagesModel: ListModel {
```
readonly (?)

noeliaSD commented on ui/app/AppLayouts/Chat/panels/communities/JoinPermissionsOverlayPanel.qml:
```diff
@@ -0,0 +1,100 @@
+            model: root.moderateHoldings
+        }
+
+        StatusButton {
```
Good catch!

noeliaSD commented on storybook/pages/JoinCommunityViewPage.qml:
```diff
@@ -0,0 +1,211 @@
+                image: d.image                
+                amISectionAdmin: d.amISectionAdmin
+                openCreateChat: false
+                communityItemsModel: d.communityItemsModel
```
You are right but I decided to keep it now only in `storybook` and take the decision when it is integrated since maybe we can decide to load some "real" model instead. WDYT?

micieslak replied:
IMO there is no chance to have access to real data at this stage. It would be a big problem bc users could easily change source code and "unblock" access to resources protected by our privileges system. Protocol should really protect access, it's not only matter of UI. We could as @John-44 to clarify that.

noeliaSD replied:
Well, at this point, what I'm thinking of as "real" data is just having a model that contains the names of the channels the community has, nothing else, and maybe... the color of the community. Nothing that could be vulnerable. Only something that makes the UI more interactive.

noeliaSD commented on ui/app/AppLayouts/Chat/panels/communities/JoinPermissionsOverlayPanel.qml:
```diff
@@ -0,0 +1,100 @@
+        readonly property string communityRequirementsNotMetText: qsTr("Membership requirements not met")
+        readonly property string communityRevealAddressText: qsTr("Reveal your address to join")
+        readonly property string channelRequirementsNotMetText: qsTr("Channel requirements not met")
+        readonly property string channelRevealAddressText: qsTr("Reveal your address to enter")
```
Use case added!

noeliaSD commented on ui/app/AppLayouts/Chat/panels/communities/JoinPermissionsOverlayPanel.qml:
```diff
@@ -0,0 +1,100 @@
+import QtQuick 2.14
```
I didn't realize about this second  case. In the majority of the designs the `or` breaks line. I guess this second example is just only a design issue but better to have a confirmation from @xAlisher !

micieslak commented on storybook/pages/JoinPermissionsOverlayPanelPage.qml:
```diff
@@ -0,0 +1,163 @@
+import AppLayouts.Chat.panels.communities 1.0
+
+
+import utils 1.0
```
is it necessary?

micieslak commented on storybook/pages/JoinPermissionsOverlayPanelPage.qml:
```diff
@@ -0,0 +1,163 @@
+        orientation: Qt.Vertical
+        SplitView.fillWidth: true
+
+
```
too many empty lines

micieslak commented on storybook/pages/JoinPermissionsOverlayPanelPage.qml:
```diff
@@ -0,0 +1,163 @@
+            SplitView.fillHeight: true
+
+            Rectangle {
+                id: rect
```
please keep empty line after `id` consistently

micieslak commented on ui/app/AppLayouts/Chat/panels/communities/CommunityColumnHeaderPanel.qml:
```diff
@@ -0,0 +1,59 @@
+Control {
+    id: root
+
+    property string name
```
please use aliases whenever possible

micieslak commented on ui/app/AppLayouts/Chat/panels/communities/HoldingsListPanel.qml:
```diff
@@ -0,0 +1,96 @@
+Control {
+    id: root
+
+    property var model
```
please add comment what roles are expected for that model and it's submodels

micieslak commented on ui/app/AppLayouts/Chat/panels/communities/JoinPermissionsOverlayPanel.qml:
```diff
@@ -0,0 +1,113 @@
+import AppLayouts.Chat.helpers 1.0
+import AppLayouts.Chat.controls.community 1.0
+
+import SortFilterProxyModel 0.2
```
not used

micieslak commented on ui/app/AppLayouts/Chat/panels/communities/JoinPermissionsOverlayPanel.qml:
```diff
@@ -0,0 +1,113 @@
+    property bool isInvitationPending: false
+    property bool isJoinRequestRejected: false
+    property string communityName
+    property var communityHoldings
```
usually for models we use names like `communityHoldingsModel` for readability, wdyt about using this convention here as well?

micieslak commented on ui/app/AppLayouts/Chat/views/communities/JoinCommunityView.qml:
```diff
@@ -0,0 +1,280 @@
+                    }
+                }
+
+                ListView {
```
please set `interactive: false` to disable scrolling

micieslak commented on ui/app/AppLayouts/Chat/views/communities/JoinCommunityView.qml:
```diff
@@ -0,0 +1,280 @@
+    QtObject {
+        id: d
+
+        readonly property int blurryRadius: 32
```
imo, while `blurryRadius` is used only once it could be inlined and `d` obj removed

micieslak commented on ui/app/AppLayouts/Chat/views/communities/JoinCommunityView.qml:
```diff
@@ -0,0 +1,280 @@
+    property bool requirementsMet: true
+    property bool isInvitationPending: false
+    property bool isJoinRequestRejected: false
+    property var communityHoldings
```
wdyt about adding *Model suffix?

micieslak commented on ui/app/AppLayouts/Chat/views/communities/JoinCommunityView.qml:
```diff
@@ -0,0 +1,280 @@
+                        id: overlayPannel
+
+                        anchors.centerIn: parent
+                        Layout.maximumHeight: parent.height
```
looks like leftover, right?

micieslak reviewed (changes requested):
Mainly very smal minor issues, but several of them are imo worth fixing. Anyway, great job!

noeliaSD commented on storybook/pages/JoinCommunityViewPage.qml:
```diff
@@ -0,0 +1,220 @@
+                }
+
+                JoinCommunityPermissionsEditor {
+                    channelName: d.chanelName
```
Here we have different editors at the same page so, I prefer to keep `d` and initialize everything in the same point.

micieslak reviewed (approved):
Great job!

glitchminer commented:
### QA

Hi @noeliaSD , here are the results:

---
1. OPEN Button text is not exact per Figma
Expected:
![image](https://user-images.githubusercontent.com/50769329/217650663-01d249dc-ca16-459f-bac1-f162240b7a59.png)

---
2. OPEN Disabled button is missing when requirements not met
Expected:
![image](https://user-images.githubusercontent.com/50769329/217650734-6172a090-f7df-40a2-a3fa-7557db684b85.png)

---
3. OPEN Unexpected scroll bar appears on hover
Expected:
![image](https://user-images.githubusercontent.com/50769329/217650933-4b14d5ba-d4cf-4f89-91d1-229e0d98057a.png)

glitchminer reviewed (commented):
Reviewed: https://github.com/status-im/status-desktop/pull/9405#issuecomment-1423243578

noeliaSD commented:
> ### QA
> Hi @noeliaSD , here are the results:
> 
> 1. OPEN Button text is not exact per Figma
>    Expected:
>    ![image](https://user-images.githubusercontent.com/50769329/217650663-01d249dc-ca16-459f-bac1-f162240b7a59.png)
> 
Added case and checkbox to test in storybook!

> 2. OPEN Disabled button is missing when requirements not met
>    Expected:
>    ![image](https://user-images.githubusercontent.com/50769329/217650734-6172a090-f7df-40a2-a3fa-7557db684b85.png)
> 

Cannot reproduce it.

> 3. OPEN Unexpected scroll bar appears on hover
>    Expected:
>    ![image](https://user-images.githubusercontent.com/50769329/217650933-4b14d5ba-d4cf-4f89-91d1-229e0d98057a.png)

Fixed!

glitchminer reviewed (approved):
Thanks @noeliaSD , tested, approved.
